core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java

```diff
@@ -397,12 +414,40 @@ public SyncStatus sync() throws AccessControlException, InvalidPathException {
       DefaultFileSystemMaster.Metrics.INODE_SYNC_STREAM_SKIPPED.inc();
       return SyncStatus.NOT_NEEDED;
     }
+    if (mAvoidConcurrentSync) {
+      /*
+       * e.g.
+       * 1. [TS=100] the sync() method is called by thread A
+       * 2. [TS=110] the sync() method is called by thread B, thread B is blocked by the new introduced MetadataSyncLock
+       * 3. [TS=180] thread A finishes the metadata sync, update the SyncPathCache, setting the last sync timestamp to 180.
+       * 4. [TS=182] thread B acquired the lock and can start sync
+       * 5. [TS=182] thread B need to check if in the past (182-110=72) unit of time, the same path has been synced by other thread,
+       * if yes, just sync the sync.
+       * In case that the current ts = sync start ts so that the sync interval becomes 0,
+       * we plus an additional millisecond here.
+       */
+      // In case that current ts = sync start ts
+      long syncIntervalToPreventConcurrentSync =
+          Instant.now().toEpochMilli() - syncStartTs + 1;
+      // TODO(elega) if a previous concurrent sync failed, the sync will trigger again
+      // This can be fixed by adding a lastSyncAttemptTs in the UfsSyncPathCache for entries
+      // and let InodeSyncStream report failed metadata sync event.
+      boolean shouldSync = mUfsSyncPathCache.shouldSyncPath(mRootScheme.getPath().toString(),
```

Comment:
@tcrain what do you think? 
Essentially for the second shouldSyncPath to prevent concurrent sync, we should check the last sync attempted timestamp instead of the last succeeded sync timestamp. So that if the previous sync failed, we can skip the sync.

Reply:
I'm ok with retrying a sync immediately if it fails for consistency reasons.

Reply:
k thanks